src/RoyalLibrary/SortingExtensions.cs

```diff
@@ -5,140 +5,69 @@
-      {
-        foreach (var LastName in tempList)
-        {
-          //get the firstName(s) from tempDic by lastName
-
-          var dicfirstName = tempDic.AsParallel().First(x => x.Value == LastName).Key;
-          await Task.Delay(0);//fake await to run everything smoothly
-                              //combine names again with sorted
-          tempResult.Add(dicfirstName + LastName);
-
+          orderedList.Add(lastName, firstName);
         }
       });
-      tempDic = null;
-      tempList = null;
-      return tempResult;
-
-
-
-
 
+      return orderedList.AsParallel()
+        .AsOrdered()
+        .Select((fullName) => fullName.Value + fullName.Key).ToList();
     }
+
     /// <summary>
     /// returns a new sorted ICollection 
     /// </summary>
     /// <param name="source"></param>
     /// <param name="action"></param>
     /// <returns>An Enumerable sorted collection by LastName</returns>
-    public async static Task<ICollection<string>> RoyalSortLastAsync(this IEnumerable<string> source, Action<string> action)
+    public async static Task<IEnumerable<string>> SortByLastNameAsync(this IEnumerable<string> source, Action<string> action)
     {
-      int rowsCount = source.Count();
-      byte lastspaceIndex = 0;
-      string firstName, lastName = null;
-      var tempDic = new MultiKeysValue() { Capacity = rowsCount };
-      var tempList = new List<string>(rowsCount);
-      var tempResult = new List<string>(rowsCount);
-
-      await Task.Run(() =>
-      {
-        foreach (var item in source)
-        {
-
-          var spaces = item.Count(char.IsWhiteSpace);
-          lastspaceIndex = (byte)item.IndexOfNth(' ', spaces - 1);
-
-          //get the whole firstname(s) , then the lastName, add them to tempDic which will be used
-          //to get firstname(s) by lastName
-          //lastname will be added to the temp list and sorted everytime 
-          firstName = item.Substring(0, lastspaceIndex);
-          var lastNameLength = item.Length - firstName.Length;
-          lastName = item.Substring(lastspaceIndex, lastNameLength);
-
-          tempDic.Add(firstName, lastName);
-          tempList.Add(lastName);
-          tempList.Sort();
-
-
-        }
-      }).ContinueWith(async (filtering) =>
-      {
-        foreach (var LastName in tempList)
-        {
-          //get the firstName from tempDic by lastName
-
-          var dicfirstName = tempDic.AsParallel().First(x => x.Value == LastName).Key;
-          await Task.Delay(0);//fake await to run everything smoothly
-                              //combine names again with sorted
-          tempResult.Add(dicfirstName + LastName);
-
-        }
-      });
-      tempDic = null;
-      tempList = null;
-
-      foreach (var item in tempResult.AsParallel())
-      {
-        action.Invoke(item);
-      }
-
-      return tempResult;
-
-
-
-
-
+      var sortedList = await source.SortByLastNameAsync();
+      sortedList.AsParallel().ForAll(action);

```

Comment:
Amazing refactor , very less code and more clean 

but here in this line after 3 times of testing it fails! 

To solve this : `sortedList.AsParallel().ForAll(action);`
use `sortedList.ForEach(action);` instead - ForEach - is an extension method you coded
I tried this more than ten times and works well

Reply:
Hi, @ebradim thanks for reviewing the code. I made the changes based on your recommendations